core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java

```diff
@@ -45,14 +59,123 @@ public SqlCoalesceFunction() {
     // strategies are used.
     super("COALESCE",
         SqlKind.COALESCE,
-        ReturnTypes.LEAST_RESTRICTIVE.andThen(SqlTypeTransforms.LEAST_NULLABLE),
         null,
-        OperandTypes.SAME_VARIADIC,
+        InferTypes.RETURN_TYPE,
+        null,
         SqlFunctionCategory.SYSTEM);
   }
 
   //~ Methods ----------------------------------------------------------------
 
+  @Override public RelDataType deriveType(
+      SqlValidator validator,
+      SqlValidatorScope scope,
+      SqlCall call) {
+    // Do not try to derive the types of the operands. We will do that
+    // later, top down, when we infer the return type
+    return validateOperands(validator, scope, call);
+  }
+
+  @Override public RelDataType inferReturnType(
+      SqlOperatorBinding opBinding) {
+    // REVIEW jvs 4-June-2005:  can't these be unified?
+    if (!(opBinding instanceof SqlCallBinding)) {
+      return inferTypeFromOperands(opBinding);
+    }
+    return inferTypeFromValidator((SqlCallBinding) opBinding);
+  }
+
+  private static RelDataType inferTypeFromValidator(
+      SqlCallBinding callBinding) {
+    SqlCall coalesceCall = callBinding.getCall();
+    SqlNodeList arglist = new SqlNodeList(coalesceCall.getOperandList(),
+        coalesceCall.getParserPosition());
+    ArrayList<SqlNode> nullList = new ArrayList<>();
+    List<RelDataType> argTypes = new ArrayList<>();
+
+    final RelDataTypeFactory typeFactory = callBinding.getTypeFactory();
+
+
+    // Coalesce can only return null if all the input argument types are nullable
```

Comment:
This comment was unintuitive to me at first (but definitely correct). Can you add an example of why for posterity?